### PR TITLE
IfOp

### PR DESCRIPTION
**Pull request by Pangoraw**
- region inlining if predicate is known
 - transform to select for values which are not defined in inner regions
 - forward
 - reverse

I don't think the reverse could be used as is in a XLA pipeline since Enzyme is not yet able to remove all enzyme ops (which I think it technically could if there is no loop?).

### CODE REVIEW

File: test/lit_tests/batchtests/autodiff.mlir [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  func.func @f(%arg0: tensor<f32>) -> tensor<f32> {
+    %0 = stablehlo.multiply %arg0, %arg0 : tensor<f32>
+    return %0 : tensor<f32>
+  }
+
+  func.func @main(%arg0: tensor<10xf32>) -> tensor<10xf32> {
+    %0 = enzyme.batch @f(%arg0) { batch_shape = array<i64: 10> } : (tensor<10xf32>) -> tensor<10xf32>
+    return %0 : tensor<10xf32>
+  }
+}
```
Comment (wsmoses) [inline test/lit_tests/batchtests/autodiff.mlir:11]: can you add the test checks?

File: src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp [hunk 2]
```diff
@@ -6311,6 +6311,76 @@ struct IfInline final : OpRewritePattern<mlir::stablehlo::IfOp> {
...
+
+    rewriter.setInsertionPointToEnd(&replacement.getFalseBranch().front());
+    rewriter.replaceOpWithNewOp<mlir::stablehlo::ReturnOp>(
+        replacement.getFalseBranch().front().getTerminator(), falseReturns);
+
+    rewriter.replaceOp(op, results);
+    return success();
+  }
+};
+
```
Comment (wsmoses) [inline src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp:6383]: Can you also add the pattern if the condition is a constant?

File: test/lit_tests/batchtests/autodiff.mlir (hunk 1 above)
Comment (Pangoraw) [reply test/lit_tests/batchtests/autodiff.mlir]: ah sorry I did not mean to include this in this pr. I will open up another PR that bumps Enzyme + JAX with this test.

File: src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp (hunk 2 above)
Comment (Pangoraw) [reply src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp:6383]: If the condition is constant then the if is removed, leaving only the content of the corresponding region due to:

https://github.com/EnzymeAD/Enzyme-JAX/blob/d91410b4a0bd41ed19c70b8b0486af190408a6e3/src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp#L6274-L6281

**Review by wsmoses — APPROVED** [review]
lgtm [though also we'll need to figure out a good interface for the push/pop removal]

**Comment by Pangoraw** [comment]
> though also we'll need to figure out a good interface for the push/pop removal

Yes. I also took a look at BatchOpInterface for IfOp and in its current state the interface only allows to return a single operation so we could not really unroll the batching like we discussed. So maybe `BatchOpInterface` has to be modified to provide a `IRBuilder` to `createBatch` and implementers do the result mapping themselves.

**Comment by wsmoses** [comment]
> > though also we'll need to figure out a good interface for the push/pop removal
> 
> Yes. I also took a look at BatchOpInterface for IfOp and in its current state the interface only allows to return a single operation so we could not really unroll the batching like we discussed. So maybe `BatchOpInterface` has to be modified to provide a `IRBuilder` to `createBatch` and implementers do the result mapping themselves.

Yeah I think so, as is it assumed a single op would suffice for batching

**Pull request merged**